sdk/nodejs/src/eventQueueAS.ts

```diff
@@ -0,0 +1,129 @@
+import { DVCEvent } from './types'
+import { DVCRequestEvent } from './models/requestEvent'
+import { DVCPopulatedUser } from './models/populatedUser'
+import { DVCLogger } from '@devcycle/types'
+
+import { getBucketingLib } from './bucketing'
+import { EventQueueInterface } from './eventQueue'
+import { publishEvents } from './request'
+
+export const AggregateEventTypes: Record<string, string> = {
+    variableEvaluated: 'variableEvaluated',
+    variableDefaulted: 'variableDefaulted',
+}
+
+export const EventTypes: Record<string, string> = { ...AggregateEventTypes }
+
+type UserEventsBatchRecord = {
+    user: DVCPopulatedUser,
+    events: DVCRequestEvent[]
+}
+export type FlushPayload = {
+    payloadId: string,
+    records: UserEventsBatchRecord[]
+}
+
+type options = {
+    flushEventsMS?: number,
+    disableAutomaticEventLogging?: boolean,
+    disableCustomEventLogging?: boolean
+}
+
+export class EventQueueAS implements EventQueueInterface {
+    private readonly logger: DVCLogger
+    private readonly environmentKey: string
+    flushEventsMS: number
+    disableAutomaticEventLogging: boolean
+    disableCustomEventLogging: boolean
+    private flushInterval: NodeJS.Timer
+
+    constructor(logger: DVCLogger, environmentKey: string, options?: options) {
+        this.logger = logger
+        this.environmentKey = environmentKey
+        this.flushEventsMS = options?.flushEventsMS || 10 * 1000
+        this.disableAutomaticEventLogging = options?.disableAutomaticEventLogging || false
+        this.disableCustomEventLogging = options?.disableCustomEventLogging || false
+
+        this.flushInterval = setInterval(this.flushEvents.bind(this), this.flushEventsMS)
+
+        getBucketingLib().initEventQueue(environmentKey, JSON.stringify(options))
+    }
+
+    cleanup(): void {
+        clearInterval(this.flushInterval)
+    }
+
+    /**
+     * Flush events in queue to DevCycle Events API. Requeue events if flush fails
+     */
+    async flushEvents(): Promise<void> {
+        const flushPayloadsStr = getBucketingLib().flushEventQueue(this.environmentKey)
+        this.logger.debug(`AS Flush Payloads: ${flushPayloadsStr}`)
+        const flushPayloads = JSON.parse(flushPayloadsStr) as FlushPayload[]
+        if (flushPayloads.length === 0) return
+
+        const innerReducer = (val: number, batch: UserEventsBatchRecord) => val + batch.events.length
+        const reducer = (val: number, batches: FlushPayload) => val + batches.records.reduce(innerReducer, 0)
+        const eventCount = flushPayloads.reduce(reducer, 0)
```

Comment:
this seems kinda expensive to do every time just for logs (copied this from nodeJS)...

Reply:
Seems like the FlushPayload could at least have a field on it for the number of events in the payload, so that you wouldn't need this inner reducer? It could just be calculated when the FlushPayload is created.

Also, later on we check if logging is enabled or not. If logging is disabled, then we shouldn't be doing any of this either?

Reply:
yea makes sense